apps/src/lib/levelbuilder/code-docs-editor/ProgrammingEnvironmentEditor.jsx

```diff
@@ -0,0 +1,164 @@
+import React, {useState} from 'react';
+import PropTypes from 'prop-types';
+import $ from 'jquery';
+import color from '@cdo/apps/util/color';
+import TextareaWithMarkdownPreview from '@cdo/apps/lib/levelbuilder/TextareaWithMarkdownPreview';
+import Button from '@cdo/apps/templates/Button';
+import UploadImageDialog from '@cdo/apps/lib/levelbuilder/lesson-editor/UploadImageDialog';
+import SaveBar from '@cdo/apps/lib/levelbuilder/SaveBar';
+import {navigateToHref} from '@cdo/apps/utils';
+
+const EDITOR_TYPES = ['blockly', 'droplet', 'text'];
+
+const useProgrammingEnvironment = initialProgrammingEnvironment => {
+  const [programmingEnvironment, setProgrammingEnvironment] = useState(
+    initialProgrammingEnvironment
+  );
+  const updateProgrammingEnvironment = (key, value) => {
+    setProgrammingEnvironment({...programmingEnvironment, [key]: value});
+  };
+
+  return [programmingEnvironment, updateProgrammingEnvironment];
+};
+
+export default function ProgrammingEnvironmentEditor({
+  initialProgrammingEnvironment
+}) {
+  const {
+    name,
+    ...remainingProgrammingEnvironment
+  } = initialProgrammingEnvironment;
+  const [
+    programmingEnvironment,
+    updateProgrammingEnvironment
+  ] = useProgrammingEnvironment(remainingProgrammingEnvironment);
+  const [uploadImageDialogOpen, setUploadImageDialogOpen] = useState(false);
+  const [isSaving, setIsSaving] = useState(false);
+  const [lastUpdated, setLastUpdated] = useState(null);
+  const [error, setError] = useState(null);
+
+  const save = () => {
+    if (isSaving) {
+      return;
+    }
+    setIsSaving(true);
+    fetch(`/programming_environments/${name}`, {
```

Comment:
can you remind me is fetch better than ajax?

Reply:
That's my understanding based on [this thread](https://codedotorg.slack.com/archives/C0T0PNR0D/p1619714179041200). I don't have much knowledge here as to _why_ though